### PR TITLE
Use env vars instead of hardcoding URLs

### LINKED ISSUE #286: Use Sites instead of hard-coding URLs
_Originally posted by @dchiquito in https://github.com/dandi/dandi-api/pull/279#discussion_r631981359_

### PR DESCRIPTION
**Pull request by mvandenburgh**
Closes #286 and #612.

~~Requires https://github.com/dandi/dandischema/pull/98 and a new schema release before merging.~~ Edit: we're leaving any schema changes for later, for now just leave the hardcoded URLs in the metadata

Requires https://github.com/dandi/dandi-cli/pull/845 and https://github.com/dandi/dandi-infrastructure/pull/104 (plus a Terraform deploy)

The original issue proposed using Django's built in Sites framework, but there's a few issues with that:

1) Assuming I am reading the docs correctly (https://docs.djangoproject.com/en/3.2/ref/contrib/sites/), Django Sites is intended for the use case where you have a single Django server deployment that is pointed at by multiple URLs. Our use case is slightly different, in that we have multiple URLs that each point to a different Django server deployment
2) Django Sites is designed for websites that are entirely powered by Django, and is only applicable to Django deployments. So, using Sites would let us remove all references to api.dandiarchive.org/api-staging.dandiarchive.org but wouldn't cover the URLs to the frontend Vue app (gui.dandiarchive.org / gui-staging.dandiarchive.org).
3) Django Sites is entirely configured through the database. So, we'd either have to write a custom migration to update the sites that somehow runs differently on production vs staging (also not ideal since we'd be hardcoding deployment details), or we'd need to manually update the `Sites` page in the Django admin. 

So, I decided to instead make the web app and Django server URLs environment variables. We'd have to do this anyways for the web app even if we used Sites for the Django server, plus using environment variables allows us to set everything through Terraform and not have to touch the DB. 

If everyone is happy with these changes, I'll open a PR in dandi-infrastructure to set these variables.

### CODE REVIEW

**Review by waxlamp — COMMENTED** [review]
What would be our production deployment values for `DANDI_WEB_APP_URL` and `DANDI_API_URL`?

File: dandiapi/api/mail.py [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
 from django.core import mail
-from django.template.loader import render_to_string
+from django.template.loader import render_to_string as _render_to_string
 
 logger = logging.getLogger(__name__)
 
 
+def render_to_string(template: str, render_context: Dict[str, Any]):
+    render_context['dandi_api_url'] = settings.DANDI_API_URL
+    render_context['dandi_gui_url'] = settings.DANDI_WEB_APP_URL
```
Comment (waxlamp) [inline dandiapi/api/mail.py:15]: What is this `render_context` object used for, and why does it have a different name for what I would expect to be `dandi_web_app_url`?

Comment (mvandenburgh) [reply dandiapi/api/mail.py]: `render_context` is a dictionary of variables passed to the Django templating engine that provides the variables the templates use. Anytime you see `{{ <variable_name> }}` in a template, it's getting that variable from the render context object.

Comment (mvandenburgh) [reply dandiapi/api/mail.py]: > What is this `render_context` object used for, and why does it have a different name for what I would expect to be `dandi_web_app_url`?

Yeah it should be `dandi_web_app_url`, I originally had the Django setting called `DANDI_GUI_URL` but I renamed it and forgot to update the context variables. It's fixed now.

**Comment by mvandenburgh** [comment]
> What would be our production deployment values for `DANDI_WEB_APP_URL` and `DANDI_API_URL`?

For production we would have `DANDI_WEB_APP_URL` set to `https://gui.dandiarchive.org` and `DANDI_API_URL` set to `https://api.dandiarchive.org`

**Comment by mvandenburgh** [comment]
CI is failing because the metadata validation expects a hardcoded `dandiarchive.org` URL to be in the metadata, but in the test environment it's set to `localhost:8085`. I'll open a PR upstream in dandischema.

**Comment by waxlamp** [comment]
> > What would be our production deployment values for `DANDI_WEB_APP_URL` and `DANDI_API_URL`?
> 
> For production we would have `DANDI_WEB_APP_URL` set to `https://gui.dandiarchive.org` and `DANDI_API_URL` set to `https://api.dandiarchive.org`

This poses a slight problem. We're trying to get away from the `gui` subdomain, so it would be counterproductive to sprinkle those URLs through the help text and in other places. Will it work properly to set that to `https://dandiarchive.org` instead?

**Comment by mvandenburgh** [comment]
> > > What would be our production deployment values for `DANDI_WEB_APP_URL` and `DANDI_API_URL`?
> > 
> > 
> > For production we would have `DANDI_WEB_APP_URL` set to `https://gui.dandiarchive.org` and `DANDI_API_URL` set to `https://api.dandiarchive.org`
> 
> This poses a slight problem. We're trying to get away from the `gui` subdomain, so it would be counterproductive to sprinkle those URLs through the help text and in other places. Will it work properly to set that to `https://dandiarchive.org` instead?

Actually I misspoke, `DANDI_WEB_APP_URL` would be `https://dandiarchive.org/` (no www) since that is the hardcoded URL currently being used in production.

**Review by waxlamp — APPROVED** [review]

File: dandiapi/api/mail.py [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
-from django.template.loader import render_to_string
+from django.template.loader import render_to_string as _render_to_string
 
 logger = logging.getLogger(__name__)
 
 
+def render_to_string(template: str, render_context: Dict[str, Any]):
+    render_context['dandi_api_url'] = settings.DANDI_API_URL
+    render_context['dandi_web_app_url'] = settings.DANDI_WEB_APP_URL
+    return _render_to_string(template, render_context)
```
Comment (dchiquito) [inline dandiapi/api/mail.py:16]:
```suggestion
from django.template.loader import render_to_string

logger = logging.getLogger(__name__)


BASE_RENDER_CONTEXT = {
    'dandi_api_url': settings.DANDI_API_URL,
    'dandi_web_app_url': settings.DANDI_WEB_APP_URL,
}
```

Then you can just do something like this whenever you need to customize the render context:
```
render_context = {
    **BASE_RENDER_CONTEXT,
    'foo': 'bar',
}
```
That reads easier to me than doing function composition, but the difference is purely cosmetic, so your call.

File: dandiapi/api/views/auth.py [hunk 3]
```diff
@@ -96,5 +97,9 @@ def user_questionnaire_form_view(request: HttpRequest) -> HttpResponse:
     return render(
         request,
         'api/account/questionnaire_form.html',
-        {'questions': QUESTIONS, 'query_params': request.GET.dict()},
+        {
+            'questions': QUESTIONS,
+            'query_params': request.GET.dict(),
+            'dandi_web_app_url': settings.DANDI_WEB_APP_URL,
```
Comment (dchiquito) [inline dandiapi/api/views/auth.py:103]: Why is this included now?

**Review by dchiquito — APPROVED** [review]
Looks good to me! I had thought that the sites framework was the "correct" way to record domains, but you make a compelling argument. Just a couple of cosmetic questions.

**Comment by dchiquito** [comment]
> CI is failing because the metadata validation expects a hardcoded dandiarchive.org URL to be in the metadata, but in the test environment it's set to localhost:8085. I'll open a PR upstream in dandischema.

There's some magic already that should have dealt with that: https://github.com/dandi/dandischema/blob/master/dandischema/models.py#L41-L45

**Comment by dchiquito** [comment]
I think you're going to have to add those environment variables to the github action config and in dandi-cli.

File: dandiapi/api/views/auth.py (hunk 3 above)
Comment (mvandenburgh) [reply dandiapi/api/views/auth.py:103]: [The base template that this one extends uses it](https://github.com/dandi/dandi-api/blob/use-django-sites/dandiapi/api/templates/api/account/base.html)
